Add first draft of balenaOS masterclass (#1)
balena-io/balenaos-masterclass: README.md

balenaOS Masterclass
===========================

# Prerequisite Classes

This is the first balenaOS masterclass.

There are no prerequisites apart from basic linux knowledge and terminal access to a device running balenaOS.

**Masterclass Type:** Core **Maximum Expected Time To Complete:** 60 minutes

# Introduction

This masterclass covers some common things that are asked about quite often such as
- Filesystem layout and mount points.
- Systemd services and journalctl
- Finding free disk space
- config.json (including a nicer way to edit them at runtime)
- Editing balenaOS files (conf/systemd services) at runtime
- Making NetworkManager logs more verbose in 4 different ways
- Time/NTP/Chrony
- Some dbus examples

# Hardware and Software Requirements

- Access to any device running balenaOS version 2.20+.
-_A Raspberry Pi 3 or balenaFin would be best as a handful of examples require a wifi device_

# Exercises

Whenever 'terminal' is mentioned, this could mean
- Accessing the device terminal via the dashboard
- or accessing via `balena ssh`

*Unless explicitly stated, it means accessing the Host balenaOS and not a container*

> # Filesystem and partition layout

Reviewer: ```suggestion
# 1. Filesystem and partition layout
```

just to keep with other classes stylistically, ditto all the other headers beneath.

balenaOS uses a specific filesystem layout. There are 6 partitions:

> - `resin-boot`:Contains the boot files

Reviewer: ```suggestion
- `resin-boot`: Contains the boot files
```

|-sda3   8:3    0  312M  0 part /mnt/sysroot/active
|-sda4   8:4    0    1K  0 part
|-sda5   8:5    0   20M  0 part /mnt/state
`-sda6   8:6    0  7.3G  0 part /mnt/data
```

The `MOUNTPOINT` column above is where you can see specific files in various partitions.
e.g. the boot files are in `/mnt/boot`

We have 2 copies of the root filesystem. One is active and running. The other is for the hostOS update.
`/mnt/sysroot/active` points to which partition is currently active and balenaOS is running from.
`/mnt/sysroot/inactive` points to which partition will get the update if we update the new OS.

If you want to check the partition labels such as `resin-boot`, `resin-rootA` check `/dev/disk/by-label/`

```
root@123123:~# ls -al /dev/disk/by-label/
total 0
drwxr-xr-x 2 root root 200 Nov  5 08:06 .
drwxr-xr-x 7 root root 140 Nov  5 08:06 ..
lrwxrwxrwx 1 root root  10 Nov  5 08:06 resin-boot -> ../../sda1
lrwxrwxrwx 1 root root  10 Nov  5 08:06 resin-data -> ../../sda6
lrwxrwxrwx 1 root root  10 Nov  5 08:06 resin-rootA -> ../../sda2
lrwxrwxrwx 1 root root  10 Nov  5 08:06 resin-rootB -> ../../sda3
lrwxrwxrwx 1 root root  10 Nov  5 08:06 resin-state -> ../../sda5
root@123123:~#
```

> ### State partition and root overlay

Reviewer: Is there meant to be content here?

But I want to know more:
- `/resin-boot`: is a copy of the boot files that end up in the boot partition. The real currently running boot files are in `/mnt/boot`. These copies are part of the OS package and used to update the boot partition during a HostOS Update.

- `/boot`: These are just the containers copy. The real boot files from bootloaders perspective are in `/mnt/sysroot/active/current/boot`.

# systemd services

We use systemd as the init system in balenaOS. There are various systemd services that handle many different parts of the OS.

## List of Key systemd services and descriptions:
- `chronyd.service` : A daemon that manages time in the OS via NTP.
- `NetworkManager.service` : A daemon that manages network connections
- `ModemManager.service` : A daemon that manages 3g/4g modems connections
- `balena.service` Runs the balenaEngine(docker) daemon on the device
- `resin-supervisor.service` : Runs the balena-supervisor container
- `openvpn.service`: openVPN daemon to connect with balenaCloud VPN.

Other services that are not as commonly asked about
- `avahi-daemon.service`: avahi advertises network services on the local network.

> - `plymouth*.service`: daemon that manages the balenaOS logo on the screen when booted. splash screen

Reviewer: ```suggestion
- `plymouth*.service`: daemon that manages the balenaOS logo on the screen when booted (splash screen)
```

- `*getty*.service`: provides a login shell over serial/hdmi. Username: `root`

## Checking state of device and/or services:
`systemctl` is a cli utility as part of systemd that can be used to check various services.

> Some common uses

Reviewer: ```suggestion
Some common uses:
```

# Finding free disk space.
The data and state sometimes fill up. We have mitigations. But if they fill up, bad things happen.

`df` is the utility you are looking for.

```
root@123123:~# df -h | grep mnt
/dev/disk/by-label/resin-rootB  300M  266M   14M  96% /mnt/sysroot/active
/dev/disk/by-label/resin-state   19M  230K   17M   2% /mnt/state
/dev/sda1                        40M  2.5M   38M   7% /mnt/boot
/dev/sda6                       7.2G  295M  6.6G   5% /mnt/data
/dev/sda2                       300M  265M   16M  95% /mnt/sysroot/inactive
root@123123:~#
```

- `-h` : human readable
- `| grep mnt` : only interested in real partitions and not the virtual file systems

# NetworkManager

## Connect to a wifi ssid while running balenaOS.
Lets use `nmcli` to connect a
- Syntax: `nmcli device wifi connect SSID password 'PASSWORD'`
- Example: `nmcli device wifi connect AndroidAP_1234 password 'nopassword'`

# config.json

`config.json` is a file on the device in the boot partition that is the source of truth about lots of useful bits of information.

*The real file is `/mnt/boot/config.json`. Not `/resin-boot/config.json`*

> # pretty-printing config.json.

Reviewer: ```suggestion
## pretty-printing config.json.
```

  "listenPort": 48484,
  "localMode": true,
  "mixpanelToken": "b8a821156f510cb2c0e62d4433ea2045",
  "persistentLogging": false,
  "pubnubPublishKey": "",
  "pubnubSubscribeKey": "",
  "registryEndpoint": "registry2.balena-cloud.com",
  "userId": 37413,
  "username": "zubair",
  "uuid": "b8a821156f510cb2c0e62d4433ea2045",
  "vpnEndpoint": "vpn.balena-cloud.com",
  "vpnPort": 443,
  "registered_at": 1568979964264,
  "deviceId": 1701155
}
```

You can check a specific key using

```
root@123123:~# jq .deviceType /mnt/boot/config.json
"intel-nuc"
root@123123:~#
```

Edits to `config.json` should not generally be needed. Changing various options in the balenaCloud dashboard results in the supervisor safely editing `config.json` and restarting the specific service that consumes the specific options. Editing by hand manually is an advanced topic discussed later.

# Editing the core OS files at runtime

balenaOS root filesystem is read-only by default for more robustness. But editing the OS files can be quite useful if you want to add more logging/debugging flags while investigating an issue.

We can switch to read-write mode using the following command.
- `mount -o remount,rw /`

> This remounts the filesystem in readwrite mode. Then you can just edit files.

Reviewer: ```suggestion
This remounts the filesystem in read-write mode. Then you can just edit files.
```

- Edit `vi /etc/NetworkManager/NetworkManager.conf`
- Add the following at the end
```
[logging]
level=DEBUG
```
- Restart NM service.
 - `systemctl daemon-reload`
 - `systemctl restart NetworkManager`

#### Edit the systemd service and pass flags
Check the system service status to see the service file.
```
root@123123:~# systemctl status Netwo*
● NetworkManager.service - Network Manager
   Loaded: loaded (/lib/systemd/system/NetworkManager.service; enabled; vendor preset: enabled)
  Drop-In: /etc/systemd/system/NetworkManager.service.d
           └─NetworkManager.conf
```

Need to edit either the Loaded file or the Drop-in file. Depending on where `ExecStart=` is located.
If `ExecStart=` exists in Drop-In: file, it will take precedence.

- Edit `/lib/systemd/system/NetworkManager.service` using `vi`
- Find the `ExecStart=` line
- Append `--log-level=INFO`
- Restart NM service.
 - `systemctl daemon-reload`
 - `systemctl restart NetworkManager`

# Check kernel messages

Use `dmesg` to see the kernel messages

# Check if a kernel config option is enabled.

A copy of the kernel configuration is always available on a device in `/proc/config.gz`

> Here is how you would search it if `CONFIG_SPI` is enabled in a dvice.

Reviewer: ```suggestion
Here is how you would search it if `CONFIG_SPI` is enabled in a device.
```

```
root@123123:~# zcat /proc/config.gz | grep -i config_spi*
...
# CONFIG_SPI is not set
```

So SPI is not enabled in the kernel for this device

# balenaOS on your laptop

> We can use docker to spin up a balenaOS container and run bash to poke around it. Useful for various use-case.s

Reviewer: ```suggestion
We can use docker to spin up a balenaOS container and run bash to poke around. Useful for various use-cases.
```

bash-4.4# mmcli --version

mmcli 1.10.6
Copyright (2011 - 2019) Aleksander Morgado
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl-2.0.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

bash-4.4#
```

## Run arm/aarch64 balenaOS images on your laptop

We can use qemu to run different arch docker images on our laptop. Windows/Mac users can use the same way as above. Linux users will need to install `binfmt-misc` and `qemu-user-static` packages and mount a statically linked qemu binary inside the container such as:

```
zubairlk@zubair-xps-resin:~/resin/yocto/balena-intel$ docker run -it --rm -v /usr/bin/qemu-aarch64-static:/usr/bin/qemu-aarch64-static -v /usr/bin/qemu-arm-static:/usr/bin/qemu-arm-static resin/resinos:2.44.0_rev1.dev-raspberry-pi /bin/bash
bash-4.4# cat /etc/os-release
ID="balena-os"
NAME="balenaOS"
VERSION="2.44.0+rev1"
VERSION_ID="2.44.0+rev1"
PRETTY_NAME="balenaOS 2.44.0+rev1"
MACHINE="raspberrypi"
VARIANT="Development"
VARIANT_ID="dev"
META_BALENA_VERSION="2.44.0"
RESIN_BOARD_REV="8fc90cc"
META_RESIN_REV="7fed82f"
SLUG="raspberry-pi"
bash-4.4#
```

> # Advanced: Editing config.json by hand.

Reviewer: To me this section is something we should be heavily discouraging (instead redirecting to https://github.com/balena-io-playground/configizer or similar), but maybe it's useful for debugging?
